### PR TITLE
DAOS-15045 vos: check tls before accessing

### PR DESCRIPTION
**Pull request by wangdi1**
Check vos tls in vos_pool_hhash_get(), since it maybe called in the system xstream.

Required-githooks: true

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate watchers.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

**Review by jolivier23 — APPROVED** [review]

**Review by liuxuezhao — APPROVED** [review]

File: src/vos/vos_tls.h [hunk 1]
```diff
@@ -74,6 +74,13 @@ struct vos_tls *vos_tls_get(bool standalone);
 static inline struct d_hash_table *
 vos_pool_hhash_get(bool is_sysdb)
 {
+#ifndef VOS_STANDALONE
+	/* main thread doesn't have TLS and XS context or service thread
+	 * is not initialized yet.
+	 */
+	if (dss_tls_get() == NULL)
+		return NULL;
+#endif
```
Comment (wangshilong) [inline src/vos/vos_tls.h:83]: This looks unnecessary to me, the comments logic has been handled in vos_tls_get() codes:
struct vos_tls *
vos_tls_get(bool standalone)
{
#ifdef VOS_STANDALONE
        return self_mode.self_tls;
#else

I think the bug here is vos_pool_hhash_get(is_sysdb); return NULL which cause core dump inside d_uhash_link_lookup()

Comment (NiuYawei) [reply src/vos/vos_tls.h:83]: Indeed, and I don't think this PR fixed the real bug.

It looks to me the real bug is that on mgmt module setup, it calls vos_pool_kill() on main thread (not any xstream) directly to cleanup leftover RDB.

**Review by NiuYawei — CHANGES_REQUESTED** [review]

File: src/vos/vos_tls.h (hunk 1 above)
Comment (NiuYawei) [reply src/vos/vos_tls.h:83]: see sg_mgmt_tgt_setup() -> cleanup_leftover_pools() -> cleanup_leftover_cb() -> clear_vos_pool() -> vos_pool_kill(). It's called on main thread on start, so we'd ensure the vos_pool_kill() is called by a ULT on sys xstream.

File: src/mgmt/srv_target.c [hunk 2]
```diff
@@ -264,15 +264,32 @@ struct dead_pool {
...
 	struct d_uuid		 id;
+	struct vos_pool_kill_arg arg;
 
 	/* destroy blobIDs */
 	D_DEBUG(DB_MGMT, "Clear SPDK blobs for pool "DF_UUID"\n", DP_UUID(uuid));
-	rc = vos_pool_kill(uuid, VOS_POF_RDB);
+	uuid_copy(arg.pool_uuid, uuid);
+	arg.flags = VOS_POF_RDB;
+
+	rc = dss_ult_execute(vos_pool_kill_ult, &arg, NULL, NULL, DSS_XS_SYS, 0, 0);
```
Comment (NiuYawei) [inline src/mgmt/srv_target.c:292]: Minor: could reuse the tgt_kill_pool() function?

File: src/vos/vos_pool.c [hunk 3]
```diff
@@ -1,5 +1,5 @@
 /**
- * (C) Copyright 2016-2023 Intel Corporation.
+ * (C) Copyright 2016-2024 Intel Corporation.
```
Comment (NiuYawei) [inline src/vos/vos_pool.c:2]: This change should be reverted.

File: src/vos/vos_tls.h [hunk 4]
```diff
@@ -1,5 +1,5 @@
 /**
- * (C) Copyright 2016-2023 Intel Corporation.
+ * (C) Copyright 2016-2024 Intel Corporation.
```
Comment (NiuYawei) [inline src/vos/vos_tls.h:2]: ditto.

**Review by NiuYawei — APPROVED** [review]

**Review by liuxuezhao — APPROVED** [review]

**Review by NiuYawei — APPROVED** [review]

**Pull request merged**